sql: TestDistSQLDrainingHosts failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/2cbfb514fed209e9e4192bd07af6baa8dd073bab

Parameters:
```
TAGS=
GOFLAGS=
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stressrace instead of stress and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
make stress TESTS=TestDistSQLDrainingHosts PKG=github.com/cockroachdb/cockroach/pkg/sql TESTTIMEOUT=5m STRESSFLAGS='-stderr=false -maxtime 20m -timeout 10m'
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=970979&tab=buildLog

```
I181017 05:13:22.403495 13489 server/node.go:723  [n?] node connected via gossip and verified as part of cluster "10c47298-0621-4f39-8209-cbe53954565c"
I181017 05:13:22.404093 13512 gossip/server.go:232  [n1] received initial cluster-verification connection from {tcp 127.0.0.1:42497}
I181017 05:13:22.405492 13489 server/node.go:426  [n?] new node allocated ID 2
I181017 05:13:22.405553 13489 gossip/gossip.go:393  [n2] NodeDescriptor set to node_id:2 address:<network_field:"tcp" address_field:"127.0.0.1:42497" > attrs:<> locality:<> ServerVersion:<major_val:2 minor_val:1 patch:0 unstable:1 > build_tag:"v2.2.0-alpha.00000000-1740-g2cbfb51" started_at:1539753202405503699 
I181017 05:13:22.405607 13489 storage/stores.go:242  [n2] read 0 node addresses from persistent storage
I181017 05:13:22.405659 13489 storage/stores.go:261  [n2] wrote 1 node addresses to persistent storage
I181017 05:13:22.405933 13524 storage/stores.go:261  [n1] wrote 1 node addresses to persistent storage
I181017 05:13:22.407238 13489 server/node.go:673  [n2] bootstrapped store [n2,s2]
I181017 05:13:22.407337 13489 server/node.go:547  [n2] node=2: started with [<no-attributes>=<in-mem>] engine(s) and attributes []
I181017 05:13:22.407513 13489 server/status/recorder.go:610  [n2] available memory from cgroups (8.0 EiB) exceeds system memory 16 GiB, using system memory
I181017 05:13:22.407542 13489 server/server.go:1822  [n2] Could not start heap profiler worker due to: directory to store profiles could not be determined
I181017 05:13:22.407590 13489 server/server.go:1529  [n2] starting https server at 127.0.0.1:46471 (use: 127.0.0.1:46471)
I181017 05:13:22.407608 13489 server/server.go:1531  [n2] starting grpc/postgres server at 127.0.0.1:42497
I181017 05:13:22.407619 13489 server/server.go:1532  [n2] advertising CockroachDB node at 127.0.0.1:42497
I181017 05:13:22.408709 13489 server/server.go:1585  [n2] done ensuring all necessary migrations have run
I181017 05:13:22.408732 13489 server/server.go:1588  [n2] serving sql connections
I181017 05:13:22.415496 13508 sql/event_log.go:126  [n1,client=127.0.0.1:57972,user=root] Event: "create_database", target: 52, info: {DatabaseName:test Statement:CREATE DATABASE IF NOT EXISTS test User:root}
I181017 05:13:22.419369 13508 sql/event_log.go:126  [n1,client=127.0.0.1:57972,user=root] Event: "create_table", target: 53, info: {TableName:test.public.nums Statement:CREATE TABLE test.public.nums (num INT) User:root}
I181017 05:13:22.423329 13508 storage/replica_command.go:300  [n1,s1,r1/1:/M{in-ax}] initiating a split of this range at key /Table/53/1/1 [r2]
I181017 05:13:22.433344 14018 util/stop/stopper.go:537  quiescing; tasks left:
1      [async] closedts-rangefeed-subscriber
1      [async] auto-upgrade
W181017 05:13:22.433411 14004 internal/client/txn.go:532  [n2,intExec=read-setting] failure aborting transaction: node unavailable; try another peer; abort caused by: node unavailable; try another peer
W181017 05:13:22.433473 14003 internal/client/txn.go:532  [n2,intExec=show-version] failure aborting transaction: node unavailable; try another peer; abort caused by: read-setting: context canceled
I181017 05:13:22.433553 13521 util/stop/stopper.go:537  quiescing; tasks left:
1      [async] closedts-rangefeed-subscriber
W181017 05:13:22.433588 13872 server/node.go:943  [n2] node=2: unable to log node_join event: log-event: node unavailable; try another peer
W181017 05:13:22.433802 14003 internal/client/txn.go:532  [n2,intExec=show-version] failure aborting transaction: node unavailable; try another peer; abort caused by: read-setting: context canceled
I181017 05:13:22.433986 14018 util/stop/stopper.go:537  quiescing; tasks left:
1      [async] auto-upgrade
I181017 05:13:22.434076 13869 server/server_update.go:65  [n2] failed attempt to upgrade cluster version, error: show-version: context canceled
--- FAIL: TestDistSQLDrainingHosts (0.25s)
	distsql_physical_planner_test.go:508: error executing '
					ALTER TABLE nums SPLIT AT VALUES (1);
					ALTER TABLE nums EXPERIMENTAL_RELOCATE VALUES (ARRAY[2], 1);
				': pq: cannot up-replicate to s2; missing gossiped StoreDescriptor

```

SHA: https://github.com/cockroachdb/cockroach/commits/04cba2800919bdcf6a8467e8da97ae44b77a9626

Parameters:
```
TAGS=
GOFLAGS=
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
make stressrace TESTS=TestDistSQLDrainingHosts PKG=github.com/cockroachdb/cockroach/pkg/sql TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=975251&tab=buildLog

```
I181019 06:04:32.465277 13798 gossip/client.go:129  [n?] started gossip client to 127.0.0.1:41231
I181019 06:04:32.465645 13675 gossip/server.go:232  [n1] received initial cluster-verification connection from {tcp 127.0.0.1:37177}
I181019 06:04:32.466232 13391 server/node.go:698  [n?] connecting to gossip network to verify cluster ID...
I181019 06:04:32.466263 13391 server/node.go:723  [n?] node connected via gossip and verified as part of cluster "b8bf8380-b880-4081-a715-72516255ea6a"
I181019 06:04:32.467309 13675 gossip/server.go:232  [n1] received initial cluster-verification connection from {tcp 127.0.0.1:37177}
I181019 06:04:32.468347 13391 server/node.go:426  [n?] new node allocated ID 2
I181019 06:04:32.468403 13391 gossip/gossip.go:393  [n2] NodeDescriptor set to node_id:2 address:<network_field:"tcp" address_field:"127.0.0.1:37177" > attrs:<> locality:<> ServerVersion:<major_val:2 minor_val:1 patch:0 unstable:1 > build_tag:"v2.2.0-alpha.00000000-1800-g04cba28" started_at:1539929072468355754 
I181019 06:04:32.468443 13391 storage/stores.go:242  [n2] read 0 node addresses from persistent storage
I181019 06:04:32.468472 13391 storage/stores.go:261  [n2] wrote 1 node addresses to persistent storage
I181019 06:04:32.469140 13458 storage/stores.go:261  [n1] wrote 1 node addresses to persistent storage
I181019 06:04:32.471953 13391 server/node.go:673  [n2] bootstrapped store [n2,s2]
I181019 06:04:32.472126 13391 server/node.go:547  [n2] node=2: started with [<no-attributes>=<in-mem>] engine(s) and attributes []
I181019 06:04:32.472431 13391 server/status/recorder.go:610  [n2] available memory from cgroups (8.0 EiB) exceeds system memory 16 GiB, using system memory
I181019 06:04:32.472466 13391 server/server.go:1822  [n2] Could not start heap profiler worker due to: directory to store profiles could not be determined
I181019 06:04:32.472546 13391 server/server.go:1529  [n2] starting https server at 127.0.0.1:32891 (use: 127.0.0.1:32891)
I181019 06:04:32.472583 13391 server/server.go:1531  [n2] starting grpc/postgres server at 127.0.0.1:37177
I181019 06:04:32.472595 13391 server/server.go:1532  [n2] advertising CockroachDB node at 127.0.0.1:37177
I181019 06:04:32.473628 13391 server/server.go:1585  [n2] done ensuring all necessary migrations have run
I181019 06:04:32.473654 13391 server/server.go:1588  [n2] serving sql connections
I181019 06:04:32.483481 13648 sql/event_log.go:126  [n1,client=127.0.0.1:33742,user=root] Event: "create_database", target: 52, info: {DatabaseName:test Statement:CREATE DATABASE IF NOT EXISTS test User:root}
I181019 06:04:32.490245 13648 sql/event_log.go:126  [n1,client=127.0.0.1:33742,user=root] Event: "create_table", target: 53, info: {TableName:test.public.nums Statement:CREATE TABLE test.public.nums (num INT) User:root}
I181019 06:04:32.499848 13648 storage/replica_command.go:300  [n1,s1,r1/1:/M{in-ax}] initiating a split of this range at key /Table/53/1/1 [r2]
I181019 06:04:32.502915 13928 server/server_update.go:65  [n2] failed attempt to upgrade cluster version, error: node 2 not running (UNKNOWN), cannot determine version
I181019 06:04:32.503119 13931 sql/event_log.go:126  [n2] Event: "node_join", target: 2, info: {Descriptor:{NodeID:2 Address:{NetworkField:tcp AddressField:127.0.0.1:37177} Attrs: Locality: ServerVersion:2.1-1 BuildTag:v2.2.0-alpha.00000000-1800-g04cba28 StartedAt:1539929072468355754 LocalityAddress:[]} ClusterID:b8bf8380-b880-4081-a715-72516255ea6a StartedAt:1539929072468355754 LastUp:1539929072468355754}
I181019 06:04:32.515062 13734 util/stop/stopper.go:537  quiescing; tasks left:
1      [async] closedts-rangefeed-subscriber
1      [async] auto-upgrade
I181019 06:04:32.515418 13733 util/stop/stopper.go:537  quiescing; tasks left:
1      [async] closedts-rangefeed-subscriber
I181019 06:04:32.515601 13734 util/stop/stopper.go:537  quiescing; tasks left:
1      [async] closedts-rangefeed-subscriber
--- FAIL: TestDistSQLDrainingHosts (0.31s)
	distsql_physical_planner_test.go:508: error executing '
					ALTER TABLE nums SPLIT AT VALUES (1);
					ALTER TABLE nums EXPERIMENTAL_RELOCATE VALUES (ARRAY[2], 1);
				': pq: cannot up-replicate to s2; missing gossiped StoreDescriptor

```

--- a/pkg/sql/distsql_physical_planner_test.go
+++ b/pkg/sql/distsql_physical_planner_test.go
@@ -503,17 +503,21 @@ func TestDistSQLDrainingHosts(t *testing.T) {
 	r := sqlutils.MakeSQLRunner(tc.ServerConn(0))
 	r.DB.SetMaxOpenConns(1)
 
-	r.Exec(t, "SET DISTSQL = ON")
 	// Force the query to be distributed.
-	r.Exec(
-		t,
-		fmt.Sprintf(`
-			ALTER TABLE nums SPLIT AT VALUES (1);
-			ALTER TABLE nums EXPERIMENTAL_RELOCATE VALUES (ARRAY[%d], 1);
-		`,
-			tc.Server(1).GetFirstStoreID(),
-		),
-	)
+	r.Exec(t, "SET DISTSQL = ON")
+
+	// Shortly after starting a cluster, the first server's StorePool may not be
+	// fully initialized and ready to do rebalancing yet, so wrap this in a
+	// SucceedsSoon.
+	testutils.SucceedsSoon(t, func() error {
+		_, err := r.DB.Exec(
+			fmt.Sprintf(`ALTER TABLE nums SPLIT AT VALUES (1);
+									 ALTER TABLE nums EXPERIMENTAL_RELOCATE VALUES (ARRAY[%d], 1);`,
+				tc.Server(1).GetFirstStoreID(),
+			),
+		)
+		return err
+	})
 
 	// Ensure that the range cache is populated (see #31235).
 	r.Exec(t, "SHOW EXPERIMENTAL_RANGES FROM TABLE nums")

